Build failure: shntool

### Nixpkgs version

- Unstable (26.05)

### Steps to reproduce

nix-build -A shntool

### Can Hydra reproduce this build failure?

Yes, Hydra can reproduce this build failure.

### Link to Hydra build job

https://hydra.nixos.org/build/318189779

### Relevant log output

```console
make[1]: Entering directory '/build/source/src'
gcc -DHAVE_CONFIG_H -I. -I../include  -I../include  -I../include -g -O2 -c core_convert.c
In file included from ../include/format-types.h:27,
                 from ../include/format.h:26,
                 from ../include/shntool.h:30,
                 from core_convert.c:19:
../include/module-types.h:48:13: error: 'bool' cannot be defined via 'typedef'
   48 | typedef int bool;
      |             ^~~~
../include/module-types.h:48:13: note: 'bool' is a keyword with '-std=c23' onwards
../include/module-types.h:48:1: warning: useless type name in empty declaration
   48 | typedef int bool;
      | ^~~~~~~
make[1]: *** [Makefile:311: core_convert.o] Error 1
make[1]: Leaving directory '/build/source/src'
make: *** [Makefile:249: all-recursive] Error 1
error: build of '/nix/store/izr4hw4c0n5czi4virn2gm6alwqx4cp7-shntool-3.0.10+git20130108.4ca41f4-1.drv' on 'ssh-ng://remotebuild@adynamia' failed: Cannot build '/nix/store/izr4hw4c0n5czi4virn2gm6alwqx4cp7-shntool-3.0.10+git20130108.4ca41f4-1.drv'.
       Reason: builder failed with exit code 2.
       Output paths:
         /nix/store/p45252zk3zxy7xfxlvizm0apcqy24frv-shntool-3.0.10+git20130108.4ca41f4-1
       Last 25 log lines:
       > install :  /nix/store/p45252zk3zxy7xfxlvizm0apcqy24frv-shntool-3.0.10+git20130108.4ca41f4-1/bin
       >
       > Running phase: buildPhase
       > build flags: SHELL=/nix/store/lw117lsr8d585xs63kx5k233impyrq7q-bash-5.3p3/bin/bash
       > Making all in man
       > make[1]: Entering directory '/build/source/man'
       > make[1]: Nothing to be done for 'all'.
       > make[1]: Leaving directory '/build/source/man'
       > Making all in src
       > make[1]: Entering directory '/build/source/src'
       > gcc -DHAVE_CONFIG_H -I. -I../include  -I../include  -I../include -g -O2 -c core_convert.c
       > In file included from ../include/format-types.h:27,
       >                  from ../include/format.h:26,
       >                  from ../include/shntool.h:30,
       >                  from core_convert.c:19:
       > ../include/module-types.h:48:13: error: 'bool' cannot be defined via 'typedef'
       >    48 | typedef int bool;
       >       |             ^~~~
       > ../include/module-types.h:48:13: note: 'bool' is a keyword with '-std=c23' onwards
       > ../include/module-types.h:48:1: warning: useless type name in empty declaration
       >    48 | typedef int bool;
       >       | ^~~~~~~
       > make[1]: *** [Makefile:311: core_convert.o] Error 1
       > make[1]: Leaving directory '/build/source/src'
       > make: *** [Makefile:249: all-recursive] Error 1
       For full logs, run:
         nix log /nix/store/izr4hw4c0n5czi4virn2gm6alwqx4cp7-shntool-3.0.10+git20130108.4ca41f4-1.drv
error: Cannot build '/nix/store/izr4hw4c0n5czi4virn2gm6alwqx4cp7-shntool-3.0.10+git20130108.4ca41f4-1.drv'.
       Reason: builder failed with exit code 1.
       Output paths:
         /nix/store/p45252zk3zxy7xfxlvizm0apcqy24frv-shntool-3.0.10+git20130108.4ca41f4-1
```

### Additional context

Related to https://github.com/NixOS/nixpkgs/issues/475479

### System metadata

 - system: `"x86_64-linux"`
 - host os: `Linux 6.12.63, NixOS, 26.05 (Yarara), 26.05.20251228.c0b0e0f`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.31.2+1`
 - nixpkgs: `/nix/store/xzl5dhddgjxp2p18h5qpkhvlkp8128j1-source`

### Notify maintainers

@jcumming 

---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%220.kind%3A+build+failure%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/by-name/sh/shntool/bool-type.patch
+++ b/pkgs/by-name/sh/shntool/bool-type.patch
@@ -1,0 +1,48 @@
+diff --git a/include/format.h b/include/format.h
+index 567993e..ab293f8 100644
+--- a/include/format.h
++++ b/include/format.h
+@@ -35,7 +35,7 @@ void tagcpy(unsigned char *,unsigned char *);
+ int tagcmp(unsigned char *,unsigned char *);
+ 
+ /* function to check if a file name is about to be clobbered, and if so, asks whether this is OK */
+-int clobber_check(char *);
++bool clobber_check(char *);
+ 
+ /* find an output format module with the given name */
+ format_module *find_format(char *);
+diff --git a/include/mode.h b/include/mode.h
+index a9c8cd0..fc13d4f 100644
+--- a/include/mode.h
++++ b/include/mode.h
+@@ -78,7 +78,7 @@ void create_output_filename(char *,char *,char *);
+ FILE *open_output_stream(char *,proc_info *);
+ 
+ /* function to determine if two filenames point to the same file */
+-int files_are_identical(char *,char *);
++bool files_are_identical(char *,char *);
+ 
+ /* function to remove a file if it exists */
+ void remove_file(char *);
+diff --git a/include/module-types.h b/include/module-types.h
+index c49c661..2d7b0a3 100644
+--- a/include/module-types.h
++++ b/include/module-types.h
+@@ -23,6 +23,8 @@
+ #ifndef __MODULE_TYPES_H__
+ #define __MODULE_TYPES_H__
+ 
++#include <stdbool.h>
++
+ #ifdef HAVE_CONFIG_H
+ #include "config.h"
+ #endif
+@@ -44,8 +46,6 @@
+ #define PATHSEPCHAR '/'
+ #endif
+ 
+-/* boolean type */
+-typedef int bool;
+ 
+ /* wtypes */
+ typedef unsigned long wlong;

--- a/pkgs/by-name/sh/shntool/function-pointer-type.patch
+++ b/pkgs/by-name/sh/shntool/function-pointer-type.patch
@@ -1,0 +1,26 @@
+diff --git a/src/core_mode.c b/src/core_mode.c
+index d67e317..21d3b9e 100644
+--- a/src/core_mode.c
++++ b/src/core_mode.c
+@@ -564,17 +564,17 @@ static int compare_ascii(const wave_info **w1,const wave_info **w2)
+ 
+ static void ascii_sort_files(wave_info **filenames, int numfiles)
+ {
+-  int (*cmpfunc) ();
++  int (*cmpfunc) (const void *, const void *);
+ 
+-  cmpfunc = compare_ascii;
++  cmpfunc = (int (*)(const void *, const void *))compare_ascii;
+   qsort(filenames,numfiles,sizeof(wave_info *),cmpfunc);
+ }
+ 
+ static void version_sort_files(wave_info **filenames,int numfiles)
+ {
+-  int (*cmpfunc) ();
++  int (*cmpfunc) (const void *, const void *);
+ 
+-  cmpfunc = compare_version;
++  cmpfunc = (int (*)(const void *, const void *))compare_version;
+   qsort(filenames,numfiles,sizeof(wave_info *),cmpfunc);
+ }
+ 

--- a/pkgs/by-name/sh/shntool/package.nix
+++ b/pkgs/by-name/sh/shntool/package.nix
@@ -17,10 +17,18 @@ stdenv.mkDerivation rec {
     sha256 = "sha256-Qn4LwVx34EhypiZDIxuveNhePigkuiICn1nBukoQf5Y=";
   };
 
+  patches = [
+    # Fix conflicts between reserved keywords `bool` (starting from C23) and custom `typedef`s.
+    ./bool-type.patch
+    # Fix implicit weak-typed pointer casting.
+    ./function-pointer-type.patch
+  ];
+
   buildInputs = [ flac ];
 
   prePatch = ''
-    patches=$(grep -v '#' ./debian/patches/series | while read patch; do echo "./debian/patches/$patch"; done | tr '\n' ' ')
+    additional_patches=$(grep -v '#' ./debian/patches/series | while read patch; do echo "./debian/patches/$patch"; done | tr '\n' ' ')
+    patches="$patches $additional_patches"
   '';
 
   meta = {

--- a/pkgs/by-name/sh/shntool/package.nix
+++ b/pkgs/by-name/sh/shntool/package.nix
@@ -5,7 +5,7 @@
   flac,
 }:
 
-stdenv.mkDerivation rec {
+stdenv.mkDerivation (finalAttrs: {
   version = "3.0.10+git20130108.4ca41f4-1";
   pname = "shntool";
 
@@ -13,7 +13,7 @@ stdenv.mkDerivation rec {
     domain = "salsa.debian.org";
     owner = "debian";
     repo = "shntool";
-    rev = "debian/${version}";
+    rev = "debian/${finalAttrs.version}";
     sha256 = "sha256-Qn4LwVx34EhypiZDIxuveNhePigkuiICn1nBukoQf5Y=";
   };
 
@@ -24,12 +24,14 @@ stdenv.mkDerivation rec {
     ./function-pointer-type.patch
   ];
 
-  buildInputs = [ flac ];
-
   prePatch = ''
     additional_patches=$(grep -v '#' ./debian/patches/series | while read patch; do echo "./debian/patches/$patch"; done | tr '\n' ' ')
     patches="$patches $additional_patches"
   '';
+
+  buildInputs = [
+    flac
+  ];
 
   meta = {
     description = "Multi-purpose WAVE data processing and reporting utility";
@@ -38,4 +40,4 @@ stdenv.mkDerivation rec {
     platforms = lib.platforms.all;
     maintainers = with lib.maintainers; [ jcumming ];
   };
-}
+})

